documentation/nearest-neighbor-search.html

```diff
@@ -0,0 +1,436 @@
+For a given image (e.g an product image shown in the product search result page) we can find  
+products which have a similar product image. 
+We can do this by fetching the SIFT encoding of the image we want to get similar images for 
+and use this tensor as input to the nearest neighbor query operator and search over the collection of SIFT encodings.</p>
+
+<p>
+We are also not interested in the corpus wide global nearest neighbors but only products where <em>in_stock</em> is <em>true</em>. 
+We express the query logic using  
+the <a href="query-language.html">YQL query language</a> and our <a href="query-api.html#http">search api request</a> becomes:</p>
+
+<pre>
+{
+  "yql": "select title, price from sources products where ([{\"targetHits\": 10}]nearestNeighbor(image_sift_encoding,query_vector_sift)) and in_stock = true;",
+  "ranking.features.query(query_vector_sift)": [
+    0.22507139604882176,
+    0.11696498718517367,
+    0.9418422036734729,
+    ...
+  ],
+  "ranking.profile": "image-similarity", 
+  "hits": 10 
+}
+</pre>
+
+<p>The YQL express our query logic using logical conjunction <em>and</em> so that our nearest neighbor search is restricted to documents where <em>in_stock</em> is true. 
+The <em>targetHits</em> is the desired number of close neighbors we want. 
+The total number of hits which is ranked by the ranking profile 
+depends on the query filters and how fast the nearest neighbor search algorithm converges. The operator might expose more than 10 hits
+to the <em>first-phase</em> ranking expression.  Since we are filtering on <em>in_stock</em> 
+one might also get 0 results if the entire product catalog is out of stock. 
+
+The <em>ranking.profile</em> parameter chooses the <em>rank-profile</em> we described earlier which simply ranks documents based on how similar they are in the SIFT 
+vector encoding space. 
+Finally, the <em>hits</em> parameter controls how many hits are 
+returned in the response. If the documents are partitioned over multiple content nodes the <em>targetHits</em> is per node, so with e.g 10 nodes we get at least
+100 nearest neighbors fully evaluated using the <em>first-phase</em> ranking function. 
+
+In our example the <em>first-phase</em> expression is simply the <em>closeness()</em> as calculated by the nearest neighbor 
+operator and hence the hits are not re-ordered but assigned the same score as calculated in the zero-phase by nearest neighbor search operator.
+
+We could re-order hits by changing the first-phase ranking function, for the following would re-rank the nearest neighbors which are in stock using 
+the price:</p>
+
+<pre>
+rank-profile image-similarity-price {
+  first-phase {
+    expression: if(isNan(attribute(price)) == 1,0.0, attribute(price)) 
+  }
+}
+</pre>
+<p>In this example the ranking profile simply ignores the closeness score assigned by the nns operator and instead re-orders the 
```

Comment:
The abbreviation "nns" has not be introduced. Consider using <em>nearestNeighbor</em> instead.